### PR TITLE
1st attempt

### PR DESCRIPTION
**Pull request by jaysomething**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -24,408 +22,153 @@
...
-    FloatType& multiply(const DoubleType& dt);
-    FloatType& divide(const DoubleType& dt);
-    // IntType args
-    FloatType& add(const IntType& it);
-    FloatType& subtract(const IntType& it);
-    FloatType& multiply(const IntType& it);
-    FloatType& divide(const IntType& it);
+	
+private:
+	float* m_value { nullptr };
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME: don't use `m_` in variable names.  Why did you change this?  You're just supposed to make it private, not rename it.

File: main.cpp [hunk 2]
```diff
@@ -24,408 +22,153 @@
...
 struct FloatType
 {
-    FloatType(float val) : value(new float(val)) {}
-    ~FloatType() { delete value; }
+    FloatType(float val = 0.0f) : m_value(new float(val)) {}
+    ~FloatType()
+	{
+		delete m_value;
+		m_value = nullptr;
+	}
```
Comment (matkatmusic) [inline main.cpp]: your previous implementations were fine. 
- [x] FIXME revert to previous implementation.

File: main.cpp [hunk 3]
```diff
@@ -24,408 +22,153 @@
...
 {
-    DoubleType() {}
-    DoubleType(double val) : value(new double(val)) {}
-    ~DoubleType() { delete value; }
+    DoubleType(double val = 0.0) : m_value(new double(val)) {}
+    ~DoubleType()
+	{
+		delete m_value;
+		m_value = nullptr;
+	}
```
Comment (matkatmusic) [inline main.cpp]: your previous implementations were fine. 
- [x] FIXME revert to previous implementation.

File: main.cpp [hunk 4]
```diff
@@ -24,408 +22,153 @@
...
-    IntType(int val) : value(new int(val)) {}
-    ~IntType() { delete value; }
-
-    int* value;
+    IntType(int val = 0) : m_value(new int(val)) {}
+    ~IntType()
+	{
+		delete m_value;
+		m_value = nullptr;
+	}
```
Comment (matkatmusic) [inline main.cpp]: your previous implementations were fine. 
- [x] FIXME revert to previous implementation.

File: main.cpp [hunk 5]
```diff
@@ -24,408 +22,153 @@
...
-    DoubleType& multiply(const DoubleType& dt);
-    DoubleType& divide(const DoubleType& dt);
-    // IntType args
-    DoubleType& add(const IntType& it);
-    DoubleType& subtract(const IntType& it);
-    DoubleType& multiply(const IntType& it);
-    DoubleType& divide(const IntType& it);
+	
+private:
+	double* m_value { nullptr };
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME: don't use `m_` in variable names.  Why did you change this?  You're just supposed to make it private, not rename it.

File: main.cpp [hunk 6]
```diff
@@ -24,408 +22,153 @@
...
-    IntType& multiply(const DoubleType& dt);
-    IntType& divide(const DoubleType& dt);
-    // IntType args
-    IntType& add(const IntType& it);
-    IntType& subtract(const IntType& it);
-    IntType& multiply(const IntType& it);
-    IntType& divide(const IntType& it);
+
+private:
+	int* m_value { nullptr };
```
Comment (matkatmusic) [inline main.cpp]: - [x] FIXME: don't use `m_` in variable names.  Why did you change this?  You're just supposed to make it private, not rename it.

File: main.cpp [hunk 7]
```diff
@@ -24,408 +22,153 @@
...
-    return *this;
-}
-
-FloatType& FloatType::divide(const IntType& it)
-{
-    std::cout << "(ft) " << *value << " / (it) " << *it.value << " = ";
-    *value = *value / *it.value;
-    std::cout << *value << '\n';
+	std::cout << "(ft) " << *m_value << " / " << v << " = " << *m_value / v << '\n';
+	*m_value /= v;
```
Comment (matkatmusic) [inline main.cpp]: - [x] remove these std::cout statements.

File: main.cpp [hunk 8]
```diff
@@ -24,408 +22,153 @@
...
-    return *this;
-}
-
-DoubleType& DoubleType::divide(const IntType& it)
-{
-    std::cout << "(dt) " << *value << " / (it) " << *it.value << " = ";
-    *value = *value / *it.value;
-    std::cout << *value << '\n';
+	std::cout << "(dt) " << *m_value << " / " << v << " = " << *m_value / v << '\n';
+	*m_value /= v;
```
Comment (matkatmusic) [inline main.cpp]: - [x] remove these std::cout statements.

File: main.cpp [hunk 9]
```diff
@@ -434,145 +177,19 @@ IntType& IntType::divide(int v)
...
-        std::cout << "Error: Int division by zero. Calculation skipped. ";
-        std::cout << "(it) " << *value << '\n';
-    }
-    else
-    {
-        std::cout << "(it) " << *value << " / (it) " << *it.value << " = ";
-        *value = *value / *it.value;
-        std::cout << *value << '\n';
+        std::cout << "(it) " << *m_value << " / " << v << " = " << *m_value / v << '\n';
+        *m_value /= v;
```
Comment (matkatmusic) [inline main.cpp]: remove these std::cout statements

File: main.cpp [hunk 10]
```diff
@@ -593,24 +210,17 @@ int main()
 
     std::cout << "- ";
     std::cout << "FloatType UDT calculations:\n";
-    ft.add(it).multiply(ft).subtract(dt).divide(it);
+    ft.add(it).multiply(ft).subtract(static_cast<float>(dt)).divide(it);
```
Comment (matkatmusic) [inline main.cpp]: - [x] revise this to say: 
```
std::cout << "FloatType UDT calculations: " << ft.add(it).multiply(ft).subtract(static_cast<float>(dt)).divide(it) << "\n";
```
This will cause the conversion functions to be invoked by the std::cout statement.

File: main.cpp [hunk 11]
```diff
@@ -593,24 +210,17 @@ int main()
...
     std::cout << "IntType UDT calculations:\n";
-    it.add(it).multiply(ft).subtract(ft).divide(dt);
-    std::cout << '\n';
-
-    *ft.value = 1.0f;
-    *dt.value = 1.0;
-    *it.value = 1;
-    std::cout << "- ";
-    std::cout << "Object values set to 1.\n";
+    it.add(it).multiply(static_cast<int>(ft)).subtract(static_cast<int>(ft)).divide(static_cast<int>(dt));
```
Comment (matkatmusic) [inline main.cpp]: - [x] see above comment.

**Review by matkatmusic — CHANGES_REQUESTED** [review]
you've got a bunch of extra std::cout statements, and you changed your variable name to a style we don't use in the course.

File: main.cpp (hunk 1 above)
Comment (jaysomething) [reply main.cpp]: Cos I saw it used all over the place on learncpp.com and thought it was good practice. 

Here's the related paragraph

> Using the “m_” prefix for member variables helps distinguish member variables from function parameters or local variables inside member functions. This is useful for several reasons. First, when we see an assignment to a variable with the “m_” prefix, we know that we are changing the state of the class instance. Second, unlike function parameters or local variables, which are declared within the function, member variables are declared in the class definition. Consequently, if we want to know how a variable with the “m_” prefix is declared, we know that we should look in the class definition instead of within the function.

So I suppose this would be an instance of the "Hungarian notation"? If so I just saw it on the JUCE coding standard that it's not allowed.

File: main.cpp (hunk 2 above)
Comment (jaysomething) [reply main.cpp]: I saw a reply on [this thread](https://stackoverflow.com/questions/1931126/is-it-good-practice-to-null-a-pointer-after-deleting-it) and again thought it was good practice, just like checking for `nullptr` before using it.

**Review by matkatmusic — APPROVED** [review]
